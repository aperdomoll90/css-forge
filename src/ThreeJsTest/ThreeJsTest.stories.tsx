import { Meta, StoryFn } from '@storybook/react'
import {ThreeJsTest} from './ThreeJsTest'
import StorybookContainer from '../StorybookContainer'
import { ThreeJsTestPropsType } from './ThreeJsTest.types'

export default {
  title: '3D',
  component: ThreeJsTest,
  parameters: {
    jest: ['ThreeJsTest.test.tsx'],
  },
  argTypes: {
    size: { options: [90, 80, 50,], control: { type: 'radio' } },
  },
} as Meta

const ThreeJsTestTemplate: StoryFn<ThreeJsTestPropsType> = (args) => (
  <StorybookContainer>
    <ThreeJsTest {...args} />
  </StorybookContainer>
)

export const ThreeJsTestComponent = ThreeJsTestTemplate.bind({})
ThreeJsTestComponent.args = {
  size: 80,
}