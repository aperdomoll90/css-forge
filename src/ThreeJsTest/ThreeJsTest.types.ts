export interface ThreeJsTestPropsType {
    size?: number;
  }