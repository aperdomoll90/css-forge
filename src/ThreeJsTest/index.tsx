export { ThreeJsTest } from "./ThreeJsTest"
export * from './ThreeJsTest.types'