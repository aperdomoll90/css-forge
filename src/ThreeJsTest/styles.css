.three-wrapper {
  position: relative;
  height: 100vh;
  width: 100vw;
  display: flex;
  justify-content: center;
  align-items: center;
  z-index: 99;
}

.fixed-square {
  position: fixed;
  top: 50%;
  left: 50%;
  width: 10rem;
  aspect-ratio: 1;
  background: red;
}

.three-scroll-area-wrapper {
  position: absolute;
  top: 0;
  left: 0;
  width: 100vw;
  height: 100vh;
  overflow: auto;
}
.three-scroll-dom-content {
  position: sticky;
  top: 0;
}

.canvas-html-container {
  margin: 0 auto;
  width: 100%;
  max-width: 100%;
  height: 100%;
  display: flex;
  align-items: center;
  justify-content: center;
}

.canvas-html-container h1 {
  color: #fff;
  font-size: 5rem;
  text-align: center;
  width: 100%;
  margin: 0 auto;
  color: #fff;
  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto', 'Oxygen', 'Ubuntu', 'Cantarell', 'Fira Sans', 'Droid Sans', 'Helvetica Neue', sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}
