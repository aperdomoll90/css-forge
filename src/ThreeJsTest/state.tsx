import { createRef } from "react";
import { ThreeStatePropsType } from "./ThreeJsTest.types";

// helps to define the hight of our entire project

const state:ThreeStatePropsType = {
  sections: 3,
  pages: 3, //returns a max of 100vh 
  zoom: 1,
  top: createRef<any>(),
};

export default state;