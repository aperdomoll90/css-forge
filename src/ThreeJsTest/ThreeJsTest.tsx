import { ThreeJsTestPropsType } from './ThreeJsTest.types'
import './styles.css'

export const ThreeJsTest: React.FC<ThreeJsTestPropsType> = ({ size }) => {
  const faeCircleSize = {
    width: `${size}px`,
    height: `${size}px`,
  }
  const topCircle = {
    width: `${size}px`,
    height: `${size}px`,
    top: `-${size && size / 2.5}px`,
  }

  return (
    <div className='ThreeJsTestContainer'>
      <div className='faeCircle' style={faeCircleSize}></div>
      <div className='faeCircle' style={faeCircleSize}></div>
      <div className='faeCircle' style={topCircle}></div>
      <p>loading...</p>
    </div>
  )
}
