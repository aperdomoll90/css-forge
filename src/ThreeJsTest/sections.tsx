// THIS FILE WAS COPIED FROM THE CODE PEN OF CREATOR OF REACT-THREE-FIBER Paul Henschel
//this is used to build our own scrolling area for a smother effect
import React, { createContext, useRef, useContext } from 'react'
import { useFrame, useThree } from 'react-three-fiber'
const lerp = require('lerp')
import state from './state'
import { SectionPropsType } from './ThreeJsTest.types'

const offsetContext = createContext(0)

const Section: React.FC<SectionPropsType> = ({ children, offset, factor, ...props }) => {
  const { offset: parentOffset, sectionHeight, aspect } = useSection()
  const ref = useRef<any>()
  offset = offset !== undefined ? offset : parentOffset
  useFrame(() => {
    const curY = ref.current.position.y
    const curTop = state.top.current / aspect
    ref.current.position.y = lerp(curY, (curTop / state.zoom) * factor, 0.1)
  })
  return (
    <offsetContext.Provider value={offset}>
      <group {...props} position={[0, -sectionHeight * offset * factor, 0]}>
        <group ref={ref}>{children}</group>
      </group>
    </offsetContext.Provider>
  )
}

function useSection() {
  const { sections, pages, zoom } = state
  const { size, viewport } = useThree()
  const offset = useContext(offsetContext)
  const viewportWidth = viewport.width
  const viewportHeight = viewport.height
  const canvasWidth = viewportWidth / zoom
  const canvasHeight = viewportHeight / zoom
  const mobile = size.width < 700
  const margin = canvasWidth * (mobile ? 0.2 : 0.1)
  const contentMaxWidth = canvasWidth * (mobile ? 0.8 : 0.6)
  const sectionHeight = canvasHeight * ((pages - 1) / (sections - 1))
  const aspect = size.height / viewportHeight
  return {
    aspect,
    viewport,
    offset,
    viewportWidth,
    viewportHeight,
    canvasWidth,
    canvasHeight,
    mobile,
    margin,
    contentMaxWidth,
    sectionHeight,
  }
}

export { Section, useSection }
