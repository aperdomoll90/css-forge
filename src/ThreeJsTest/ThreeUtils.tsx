import { useRef } from 'react'
import { Section } from './sections'
import { Html, useGLTF } from '@react-three/drei'
import { useFrame, useLoader } from 'react-three-fiber'
import { GLTFLoader } from 'three/examples/jsm/loaders/GLTFLoader'

// export const Model = props => {
//   const group = useRef()
//   const { nodes, materials } = useGLTF(chairUrl)
//   return (
//     <group ref={group} {...props} dispose={null}>
//       <mesh castShadow receiveShadow geometry={nodes.Curve007_1.geometry} material={materials['Material.001']} />
//       <mesh castShadow receiveShadow geometry={nodes.Curve007_2.geometry} material={materials['Material.002']} />
//     </group>
//   )
// }

// useGLTF.preload(chairUrl)
interface ModelProps {
  url: string
}
export const Model: React.FC<ModelProps> = ({ url }) => {
  const gltf = useLoader(GLTFLoader, url)
  return <primitive object={gltf.scene} dispose={null} />
}

interface SectionEnsembleProps {
  children: React.ReactNode
  bgColor?: string
  modelUrl: string
  positionY: number
  domContent?: React.MutableRefObject<HTMLDivElement>
}

export const SectionEnsemble: React.FC<SectionEnsembleProps> = ({ children, bgColor, modelUrl, positionY, domContent }) => {
  const ref = useRef<any>()
  useFrame(() => (ref.current.rotation.y += 0.01))
  return (
    <Section factor={1.5} offset={1}>
      <group position={[0, positionY, 0]}>
        <mesh ref={ref} position={[0, -35, 20]}>
          <Model url={modelUrl} />
        </mesh>
        <Html portal={domContent} fullscreen>
          <>{children}</>
        </Html>
      </group>
    </Section>
  )
}
