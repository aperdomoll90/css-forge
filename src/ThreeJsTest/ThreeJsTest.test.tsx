import { render } from '@testing-library/react'
import { ThreeJsTest } from './ThreeJsTest'
import { ThreeJsTestPropsType } from './ThreeJsTest.types'

describe('ThreeJsTest component', () => {
  let props: ThreeJsTestPropsType = {
    size: 10,
  }
  const renderComponent = () => render(<ThreeJsTest {...props} />)
  it('should match snapshots', () => {
    const { container } = renderComponent()
    expect(container).toMatchSnapshot()
  })
})
