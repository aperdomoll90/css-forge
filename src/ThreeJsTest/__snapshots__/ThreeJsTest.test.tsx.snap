// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThreeJsTest component should match snapshots 1`] = `
<div>
  <div
    class="ThreeJsTestContainer"
  >
    <div
      class="faeCircle"
      style="width: 10px; height: 10px;"
    />
    <div
      class="faeCircle"
      style="width: 10px; height: 10px;"
    />
    <div
      class="faeCircle"
      style="width: 10px; height: 10px; top: -4px;"
    />
    <p>
      loading...
    </p>
  </div>
</div>
`;
